File: test/uri_utils_test.dart
import 'package:background_downloader/src/uri/uri_helpers.dart';
import 'package:flutter_test/flutter_test.dart';

final testUrl = 'https://google.com';
final testUriWithFileScheme = Uri.parse('file:///test_file.txt');

void main() {
  group('pack and unpack', () {
    test('pack should pack filename and uri into a single string', () {
      const filename = 'myFile.txt';
      final uri = Uri.parse('content://com.example.app/document/123');

      final packedString = pack(filename, uri);

      expect(packedString, ':::$filename::::::${uri.toString()}:::');
    });

    test('unpack should unpack a valid packed string into filename and uri',
        () {
      const filename = 'myFile.txt';
      final uri = Uri.parse('content://com.example.app/document/123');
      final packedString = ':::$filename::::::${uri.toString()}:::';

      final (filename: unpackedFilename, uri: unpackedUri) =
          unpack(packedString);

      expect(unpackedFilename, filename);
      expect(unpackedUri, uri);
    });

    test(
        'unpack should return original string and null uri for simple filename string',
        () {
      const invalidPackedString = 'This is not a packed string';

      final (:filename, :uri) = unpack(invalidPackedString);

      expect(filename, invalidPackedString);
      expect(uri, isNull);
    });

    test('unpack should return null and a uri for simple uri string', () {
      const uriString = 'https://www.example.com/path/to/resource';

      final (:filename, :uri) = unpack(uriString);

      expect(filename, isNull);
      expect(uri.toString(), equals(uriString));
    });

    test('uriFromStringValue should return Uri for a valid Uri string', () {
      const uriString = 'https://www.example.com/path/to/resource';
      final expectedUri = Uri.parse(uriString);

      final resultUri = uriFromStringValue(uriString);

      expect(resultUri, expectedUri);
    });

    test('uriFromStringValue should return Uri from a valid packed string', () {
      const filename = 'myFile.txt';
      final uri = Uri.parse('content://com.example.app/document/123');
      final packedString = pack(filename, uri);

      final resultUri = uriFromStringValue(packedString);

      expect(resultUri, uri);
    });

    test('uriFromStringValue should return null for an invalid string', () {
      const invalidString = 'This is not a Uri or packed string';

      final resultUri = uriFromStringValue(invalidString);

      expect(resultUri, isNull);
    });

    test(
        'uriFromStringValue should return null for a packed string with invalid Uri',
        () {
      const filename = 'myFile.txt';
      const invalidUri = 'invalid';
      const packedString = ':::$filename::::::$invalidUri:::';

      final resultUri = uriFromStringValue(packedString);

      expect(resultUri, isNull);
    });
  });
}

File: android/src/androidTest/kotlin/UriUtilsTest.kt
import android.net.Uri
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.bbflight.background_downloader.UriUtils
import org.junit.Assert.*
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class UriUtilsTest {

    @Test
    fun pack_should_pack_filename_and_uri_into_a_single_string() {
        val filename = "myFile.txt"
        val uri = Uri.parse("content://com.example.app/document/123")

        val packedString = UriUtils.pack(filename, uri)

        assertEquals(":::$filename::::::$uri:::", packedString)
    }

    @Test
    fun unpack_should_unpack_a_valid_packed_string_into_filename_and_uri() {
        val filename = "myFile.txt"
        val uri = Uri.parse("content://com.example.app/document/123")
        val packedString = ":::$filename::::::$uri:::"

        val (unpackedFilename, unpackedUri) = UriUtils.unpack(packedString)

        assertEquals(filename, unpackedFilename)
        assertEquals(uri, unpackedUri)
    }

    @Test
    fun unpack_should_return_original_string_and_null_uri_for_simple_filename_string() {
        val invalidPackedString = "This is not a packed string"

        val (filename, uri) = UriUtils.unpack(invalidPackedString)

        assertEquals(invalidPackedString, filename)
        assertNull(uri)
    }

    @Test
    fun unpack_should_return_null_and_uri_for_simple_uri_string() {
        val uriString = "https://www.example.com/path/to/resource"

        val (filename, uri) = UriUtils.unpack(uriString)

        assertNull(filename)
        assertEquals(uri.toString(), uriString)
    }

    @Test
    fun uriFromStringValue_should_return_Uri_for_a_valid_Uri_string() {
        val uriString = "https://www.example.com/path/to/resource"
        val expectedUri = Uri.parse(uriString)

        val resultUri = UriUtils.uriFromStringValue(uriString)

        assertEquals(expectedUri, resultUri)
    }

    @Test
    fun uriFromStringValue_should_return_Uri_from_a_valid_packed_string() {
        val filename = "myFile.txt"
        val uri = Uri.parse("content://com.example.app/document/123")
        val packedString = UriUtils.pack(filename, uri)

        val resultUri = UriUtils.uriFromStringValue(packedString)

        assertEquals(uri, resultUri)
    }

    @Test
    fun uriFromStringValue_should_return_null_for_an_invalid_string() {
        val invalidString = "This is not a Uri or packed string"

        val resultUri = UriUtils.uriFromStringValue(invalidString)

        assertNull(resultUri)
    }

    @Test
    fun uriFromStringValue_should_return_null_for_a_packed_string_with_invalid_Uri() {
        val filename = "myFile.txt"
        val invalidUri = "invalid"
        val packedString = ":::$filename::::::$invalidUri:::"

        val resultUri = UriUtils.uriFromStringValue(packedString)

        assertNull(resultUri)
    }

    @Test
    fun containsUri_should_return_true_for_a_valid_Uri_string() {
        val uriString = "https://www.example.com/path/to/resource"
        assertTrue(UriUtils.containsUri(uriString))
    }

    @Test
    fun containsUri_should_return_true_for_a_valid_packed_Uri_string() {
        val filename = "myFile.txt"
        val uri = Uri.parse("content://com.example.app/document/123")
        val packedString = UriUtils.pack(filename, uri)

        assertTrue(UriUtils.containsUri(packedString))
    }

    @Test
    fun containsUri_should_return_false_for_an_invalid_string() {
        val invalidString = "This is not a Uri or packed string"

        assertFalse(UriUtils.containsUri(invalidString))
    }
}